<!DOCTYPE html>
<html>
<head>
	<meta charset='utf-8'>
	<meta http-equiv="X-UA-Compatible" content="chrome=1">
	<link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
	<link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen" />
	<link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen" />
	<link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print" />
	<style> 
		li {font-size: 12px} 
		pre code {font-size: 9px}
		#nav {text-align: center}
		#license {font-size: 12px} 
	</style>
	<!--[if lt IE 9]>
	<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
	<![endif]-->
	<title>DGEclust - Tutorial</title>
</head>

<body>
	<div id="container">
	<div class="inner">

	<header>
		<h1>DGEclust</h1>
		<h2>Hierarchical Non-Parametric Bayesian Clustering of Digital Expression Data</h2>
	</header>

	<section id="downloads" class="clearfix">
		<a href="https://github.com/dvav/dgeclust/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
		<a href="https://github.com/dvav/dgeclust/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
		<a href="https://github.com/dvav/dgeclust" id="view-on-github" class="button"><span>View on GitHub</span></a>
	</section>

	<hr>

	<section id="nav">
		<a href="index.html">Home</a> |
		<a href="tutorial.html"><strong>Tutorial</strong></a>
	</section>

	<hr>

	<section id="main_content">
		<h1>Installation</h1>

		<p>
			Just download <strong>DGEclust</strong> using the links at the top of the page and unzip it in a directory of your choice. You can 
			rename the unzipped directory to something more convenient, if you wish. That's it! No further installation procedure is necessary. In order to 
			be able to use the software you will need a fairly recent version of the <a href="http://www.scipy.org/" target=”_blank”>Scipy</a> 
			stack. DGEclust was developed and tested on a Mac, but it should be usable without problems on Linux machines, too. 
		</p>

		<h1>Quick start</h1>

		<h2>The data</h2>
		<p>
			For the sake of this tutorial, let's assume that the downloaded <strong>DGEclust</strong> files are located in a directory named <code>dgeclust</code>. 
			Also, let's assume that you 
			have a matrix of count data stored in the file <code>data.txt</code>. In this tutorial, we shall use an RNA-seq dataset from 
			yeast (<em>Saccharomyces cerevisiae</em>) 
			cultures by <a href="http://www.ncbi.nlm.nih.gov/pubmed/18451266?dopt=Abstract&holding=f1000,f1000m,isrctn" target="_blank">Nagalakshmi et al.</a> The dataset
			can be downloaded from <a href="http://genomebiology.com/content/supplementary/gb-2010-11-10-r106-s3.tgz">here</a>.  
		</p>
		
		<h2>Clustering the data</h2>
		<p>From your terminal, do the following:</p>
		<pre><code>$ cd path-to-dgeclust
$ bin/clust path-to-data-txt -g [[0,3],[1,2,4,5]] &amp;</code></pre>
		<p>
			Depending on the size of your data, this will take some time to finish. The above command runs a Gibbs sampler for a default 
			of 10K iterations. The output of the sampler is saved periodically in the directory <code>_clust</code>, which you can inspect
			to check the progress of your simulation, e.g. using the <code>tail</code> command with argument <code>-f</code>:
		</p>	
		<pre><code>$ tail -f _clust/pars.txt
0	10 	l;j
0	10 	l;j
0	10 	l;j
0	10 	l;j</code></pre>	
		<p>
			The first column in <code>pars.txt</code> gives you the number of iterations, the second gives you the estimated number of clusters
			at each iteration, while the remaining columns correspond to hyper-parameters, which depend on the specific distribution being used
			by the simulator (defaults to the <strong>Negative Binomial</strong> distribution).         
		</p>

		<p>
			There are more arguments that you can pass to <code>clust</code>. Type <code>bin/clust -h</code> for more details.
		</p>
		<p>
			After the end of the simulation, you can visualize your results using <strong>IPython</strong>:  
		</p>
		<pre><code>In [1]: cd path-to-dgeclust
In [2]:	from dgeclust.gibbs.results import GibbsOutput
In [3]:	res = GibbsOutput.read('_clust')
In [4]:	figure()
In [5]; subplot(3,1,1); plot(res.pars[:,0], res.pars[:,1], 'k'); xlabel('# iterations'); ylabel('# clusters')
In [6]; subplot(3,1,2); plot(res.pars[:,0], res.pars[:,[2,3]], 'k'); xlabel('# iterations'); ylabel('p1, p2')
In [7]; subplot(3,1,3); plot(res.pars[:,0], res.pars[:,[4,5]], 'k'); xlabel('# iterations'); ylabel('p3, p4')</code></pre>
		<p>
			It seems that the algorithm converges nicely after ~1000 iterations.  
		</p>

		<h2>Testing for differential expression</h2>
		<p>
			The output of <code>clust</code> can be used to test for differential expression. For this purpose, we use another
			program that comes with <strong>DGEclust</strong>, <code>pvals</code>, which computes the posterior probability that
			any two features between any two groups of samples in our dataset are differentially expressed. In a terminal, we type:  
		</p>
		<pre><code>$ bin/pvals -t0 1000
???????</code></pre>
		<p>
			The above command post-processes the contents of the directory <code>_clust</code> and returns a file, <code>_pvals.txt</code>,
			which includes a list of features with decreasing posterior probabilities of being differentially expressed between the two groups
			of samples in the yeast dataset:
		</p>
		<pre><code>$ head _pvals.txt
???????</code></pre>
		<p>
			Back in <strong>IPython</strong>, we can visualise our dataset using an <strong>RA diagram</strong>:
		</p>
		

		<!-- <h2>Hierarchical clustering of samples</h2>
		<p><em>(under construction</em>)</p>
		<h2>Hierarchical clustering of genes</h2>
		<p><em>(under construction</em>)</p>
	 -->


	</section>

	<footer>
		DGEclust is maintained by <a href="https://github.com/dvav">dvav</a><br>
		This page was generated by <a href="http://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
	</footer>        
	</div>
	</div>
</body>
</html>